ENH: Create boolean and integer ufuncs for isnan, isinf, and isfinite.

qwhelan opened the pull request:
For the following sample code:
```python
import numpy as np
arr = np.full(10**5, 0, bool)

for i in range(10000):
    np.isnan(arr)

```
We get the following call graph from `pprof`:
![old](https://user-images.githubusercontent.com/440095/52988621-d8722580-33b4-11e9-9aa4-cd619a3f626c.png)

This PR eliminates the trip through `npy_half` by providing specialized `npy_bool` implementations of `isnan`, `isinf`, and `isfinite` Given the limited values supported by `npy_bool`, we can trivially specify the result for all inputs. Doing so initially provided a ~20x speedup and now is closer to ~250x:
```
$ asv compare HEAD^ HEAD -s

Benchmarks that have improved:

       before           after         ratio
     [95db8c28]       [e861372b]
     <bool_ufunc~1>       <bool_ufunc>
-     1.23±0.01ms      5.03±0.09μs     0.00  bench_ufunc.IsNan.time_isnan('bool')
-        65.8±8μs       5.28±0.1μs     0.08  bench_ufunc.IsNan.time_isnan('int16')
-      87.9±0.7μs      5.32±0.05μs     0.06  bench_ufunc.IsNan.time_isnan('int32')
-         145±1μs      5.40±0.08μs     0.04  bench_ufunc.IsNan.time_isnan('int64')

Benchmarks that have stayed the same:

       before           after         ratio
     [95db8c28]       [e861372b]
     <bool_ufunc~1>       <bool_ufunc>
        119±0.6μs        118±0.4μs     1.00  bench_ufunc.IsNan.time_isnan('complex128')
          240±1μs          243±5μs     1.01  bench_ufunc.IsNan.time_isnan('complex256')
          106±1μs        105±0.3μs     0.99  bench_ufunc.IsNan.time_isnan('complex64')
         568±10μs          567±6μs     1.00  bench_ufunc.IsNan.time_isnan('float16')
         27.4±2μs      27.7±0.08μs     1.01  bench_ufunc.IsNan.time_isnan('float32')
       47.3±0.4μs       47.5±0.2μs     1.00  bench_ufunc.IsNan.time_isnan('float64')
          141±2μs        141±0.5μs     1.00  bench_ufunc.IsNan.time_isnan('longfloat')
```
The call graph is also greatly simplified:
![fixed](https://user-images.githubusercontent.com/440095/52988888-ebd1c080-33b5-11e9-8707-73e1812888bd.png)

eric-wieser commented:
This has come up before I think, but in the wider context of also providing integer specializations. I don't remember if there were objections, or if the pr just stalled. I'll comment with links when I find the PR (s?) I'm thinking of

qwhelan commented:
@eric-wieser Thanks, that would be appreciated if you're able to locate any prior discussions.

For additional context, here's a trivial example of how this manifests to a user (especially in a library like `pandas`):
```python
import numpy as np

bools = np.full(10**6, 1, bool)
ints = np.full(10**6, 1, int)

%timeit np.isnan(bools)
6.33 ms ± 195 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit np.isnan(ints)
864 µs ± 28 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So any user reliant on `NaN`-masking currently sees a `~7.3x` speedup by using `int`s instead of `bool`s for storing `bool` data, which encourages use of overly broad types for performance reasons.

eric-wieser commented:
I'm not sure I understand how a user would nan-mask bool or int data. Either way, it surprises me that the ints are faster than the bools.

eric-wieser commented:
Ok, this is the second time I've tried and failed to find that PR, so I think I'm going to give up.

You should be able to reuse your `isnan`, `isfinite`, and `isinf` loops for all the integer types.

If you want to extend this PR to that, one thing to watch out for is that these functions need to not return true on `np.datetime64('nat')` and `np.timedelta64('nat')`.

If you want to avoid this pain, you could at least add loops for `bBhHiI` types, and stick a TODO in about adding it for `lLqQ` types.

qwhelan commented:
@eric-wieser A common scenario would be using `pandas`, which adds a `NaN`-mask to most functions, including `all()` and `any()`:
```python
bools = pd.Series(np.full(10**6, 1, bool))
ints = pd.Series(np.full(10**6, 1, int))

%timeit bools.all()
7.63 ms ± 88.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit ints.all()
2.7 ms ± 20.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit np.all(bools.to_numpy())
51.4 µs ± 695 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit np.all(ints.to_numpy())
747 µs ± 46.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So pretty painful overhead from `NaN` masking in the `bool` case.
 
My PR https://github.com/pandas-dev/pandas/pull/25070 would try and fix the `pandas` side of things, but there's still some low-hanging fruit in the `numpy` case.

charris commented:
Needs a release note.

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -95,6 +97,28 @@
+        } \
+    } \
+    else { \
+            BASE_OUTPUT_LOOP(tout, op) \
```
Nit: If you're trying to match the unary case below, the indent should be one less on this line

qwhelan replied:
Thanks, fixed

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -95,6 +97,28 @@
+    do { \
+    /* condition allows compiler to optimize the generic macro */ \
+    if (IS_OUTPUT_CONT(tout)) { \
+        if (args[0] == args[1]) { \
```
This inner `if` is meaningless - you're ignoring the input, so there's no point checking if its the same as the output.

It's there in the unary case to allow `a = -a` to be optimized differently from `b = -a`. There's no difference between `a = 0` and `b = 0` to the optimizer.

qwhelan replied:
Thanks, removed

eric-wieser commented:
Thanks for adding the integer loops.

Can you check that the following passes?
```python
@pytest.mark.parametrize('nat', [np.datetime64('nat'), np.timedelta64('nat')])
def test_nat_is_not_finite(self, nat):
    try:
        assert not np.isfinite(nat)
    except TypeError:
        pass  # ok, just not implemented

@pytest.mark.parametrize('nat', [np.datetime64('nat'), np.timedelta64('nat')])
def test_nat_is_nan(self, nat):
    try:
        assert np.isnan(nat)
    except TypeError:
        pass  # ok, just not implemented

@pytest.mark.parametrize('nat', [np.datetime64('nat'), np.timedelta64('nat')])
def test_nat_is_no_tinf(self, nat):
    try:
        assert not np.isinf(nat)
    except TypeError:
        pass  # ok, just not implemented
```

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -1915,3 +1915,24 @@ def test_invalid_args(self):
+            pass  # ok, just not implemented
+
+    @pytest.mark.parametrize('nat', [np.datetime64('nat'), np.timedelta64('nat')])
+    def test_nat_is_no_tinf(self, nat):
```
Made a typo here, oops. At any rate, let's let these run. Sadly, I'm expecting them to fail.

qwhelan replied:
Just changed `no_tinf -> not_inf`. Looks like the suite is passing.

eric-wieser commented:
I'll build this branch locally and play around with it. Perhaps the failure mode I'm expecting doesn't exist, which would be great!

eric-wieser commented:
Needs a rebase - the new macros should move to `fast_loop_macros.h`

qwhelan committed:
ENH: Create boolean and integer ufuncs for isnan, isinf, and isfinite.

Previously, boolean values would be routed through the half implementations of
these functions, which added considerable overhead. Creating specialized
ufuncs improves performance by ~250x

Additionally, enable autovectorization of new isnan, isinf, and isfinite ufuncs.

qwhelan commented:
@eric-wieser Rebased and tests passing

eric-wieser commented on doc/release/1.17.0-notes.rst:
```diff
@@ -156,6 +156,11 @@ thereby saving a level of indentation
 ----------------------------------------------
 In some cases where ``np.interp`` would previously return ``np.nan``, it now
 returns an appropriate infinity.
+Specialized ``np.isnan``, ``np.isinf``, and ``np.isfinite`` ufuncs for bool and int types
```
Nit: bad merge here, needs an extra line above

qwhelan replied:
This is fixed

eric-wieser reviewed (commented):
Played around with this locally, and realized that only `int -> time` casting is allowed, not vice versa. This all looks great.

I do wonder if the benchmark is worth including though - will leave that decision to other reviewers.

qwhelan commented:
@eric-wieser I've removed the benchmark from this PR.

eric-wieser commented:
@charris, look good to merge?

qwhelan commented:
@charris Are there any desired changes to this approach? I have several more patches in this vein that I've held off on submitting until this is merged.

eric-wieser commented:
I'll go ahead and put this in - nothing here seems controversial

charris commented on numpy/core/src/umath/fast_loop_macros.h:
```diff
@@ -82,6 +84,29 @@
+#define OUTPUT_LOOP_FAST(tout, op) \
+    do { \
+    /* condition allows compiler to optimize the generic macro */ \
+    if (IS_OUTPUT_CONT(tout)) { \
```
Should be indented.

eric-wieser replied:
I think this matches the existing macros, sadly

qwhelan replied:
Please see https://github.com/numpy/numpy/pull/13208 - I have fixed the indentation for all macros in this file

charris commented on numpy/core/src/umath/fast_loop_macros.h:
```diff
@@ -82,6 +84,29 @@
+    OUTPUT_LOOP { \
+        tout * out = (tout *)op1; \
+        op; \
+    }
```
Blank line between macros.

qwhelan replied:
Fixed in https://github.com/numpy/numpy/pull/13208

charris commented on numpy/core/src/umath/fast_loop_macros.h:
```diff
@@ -82,6 +84,29 @@
+ */
+#define BASE_OUTPUT_LOOP(tout, op) \
+    OUTPUT_LOOP { \
+        tout * out = (tout *)op1; \
```
Can omit the space after `*`.

qwhelan replied:
Fixed in https://github.com/numpy/numpy/pull/13208

charris commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -644,6 +644,19 @@ BOOL__ones_like(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UN
+NPY_NO_EXPORT void
+BOOL_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
+{
+    OUTPUT_LOOP_FAST(npy_bool, *out = @val@);
```
Why not just pass the value and move the assignment to the macro where `out` is declared?

qwhelan replied:
It was entirely to preserve calling convention to be similar to the other macros - I've implemented your suggestion in https://github.com/numpy/numpy/pull/13208

charris commented:
Looks OK aside from some style/organization nits. The depth of the macro nesting makes it a bit hard to follow, but looks correct. The original behavior of promoting to float looks weird in truth, but no weirder than the functions being called on integer/boolean types :)

qwhelan commented:
@charris Thanks for the comments and please see https://github.com/numpy/numpy/pull/13208 for implementation

Thanks @eric-wieser for merging!
